### PR TITLE
Integrate package:flutter_lints into templates

### PR DESCRIPTION
**Pull request by goderbauer**
Part of https://github.com/flutter/flutter/issues/78432.

This PR enables the recommended set of lints for Flutter apps and packages from `package:flutter_lints` for apps/packages/plugins created via `flutter create`. It also adapts those templates to ensure that they are analyzer clean under the new set of lints.

There's a bug in the linter wrt treatment of empty rules sections that is fixed upstream, but not rolled into the framework yet: https://github.com/dart-lang/sdk/issues/58383

### CODE REVIEW

File: packages/flutter_tools/templates/app/analysis_options.yaml.tmpl [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  # section below, it can also be suppressed for a single line of code
+  # or a specific dart file by using the `// ignore: name_of_lint` and
+  # `// ignore_for_file: name_of_lint` syntax on the line or in the file
+  # producing the lint.
+  # rules:
+    # avoid_print: false  # Uncomment to disable the `avoid_print` rule
+    # prefer_single_quotes: true  # Uncomment to enable the `prefer_single_quotes` rule
+
+# Additional information about this file can be found at
+# https://dart.dev/guides/language/analysis-options.
```
Comment (mit-mit) [inline packages/flutter_tools/templates/app/analysis_options.yaml.tmpl:31]: nit: consider removing the period in the url to avoid copy/paste issues

**Review by mit-mit — COMMENTED** [review]
Just nits

File: packages/flutter_tools/templates/app/analysis_options.yaml.tmpl [hunk 2]
```diff
@@ -0,0 +1,30 @@
+# This file configures the analyzer, which statically analyzes Dart code to
+# check for errors, warnings, and lints.
+#
+# The issues identified by the analyzer are surfaced in the UI of Dart-enabled
+# IDEs (https://dart.dev/tools#ides-and-editors). The analyzer can also be
+# invoked from the command line by running `flutter analyze`.
+
+# The following line activates a set of recommended lints for Flutter apps
+# and packages designed to encourage good coding practises.
```
Comment (Hixie) [inline packages/flutter_tools/templates/app/analysis_options.yaml.tmpl:9]: nit: "practices"

File: packages/flutter_tools/templates/app/analysis_options.yaml.tmpl [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+include: package:lints/flutter.yaml
+
+# TODO(goderbauer): Uncomment "linter:" and "rules:" when https://github.com/dart-lang/linter/issues/2612 is fixed.
```
Comment (Hixie) [inline packages/flutter_tools/templates/app/analysis_options.yaml.tmpl:12]: if we're landing before this gets fixed, let's put the TODO (and maybe the commented-out code?) somewhere where it won't get stuck in people's analysis options for years.

File: packages/flutter_tools/templates/app/lib/main.dart.tmpl [hunk 4]
```diff
@@ -113,14 +115,16 @@ class _MyHomePageState extends State<MyHomePage> {
...
   }
 }
 {{/withPluginHook}}
 {{#withPluginHook}}
 class MyApp extends StatefulWidget {
+  const MyApp({Key? key}) : super(key: key);
+
   @override
   _MyAppState createState() => _MyAppState();
 }
```
Comment (Hixie) [inline packages/flutter_tools/templates/app/lib/main.dart.tmpl:130]: these changes are fantastic

File: packages/flutter_tools/templates/app/pubspec.yaml.tmpl [hunk 5]
```diff
@@ -22,6 +22,11 @@ version: 1.0.0+1
 environment:
   sdk: {{dartSdkVersionBounds}}
 
+# Dependencies specify other packages that your package needs in order to work.
+# To upgrade your package dependencies to the latest versions consider running
```
Comment (Hixie) [inline packages/flutter_tools/templates/app/pubspec.yaml.tmpl:26]: "automatically upgrade" maybe?

File: packages/flutter_tools/templates/app/pubspec.yaml.tmpl [hunk 6]
```diff
@@ -44,6 +49,12 @@ dev_dependencies:
   flutter_test:
     sdk: flutter
 
+  # The following imports a recommended set of lints designed to encourage good
+  # coding practises. Lints are configured and activated in the
```
Comment (Hixie) [inline packages/flutter_tools/templates/app/pubspec.yaml.tmpl:53]: practices

File: packages/flutter_tools/templates/app/pubspec.yaml.tmpl [hunk 7]
```diff
@@ -44,6 +49,12 @@ dev_dependencies:
...
+  # analysis_options.yaml file located at the root of your package.
+  lints:
+    path: /Users/goderbauer/dev/lints # TODO(goderbauer): Update this before submitting.
```
Comment (Hixie) [inline packages/flutter_tools/templates/app/pubspec.yaml.tmpl:56]: isn't this redundant with the analysis options? i'm not sure i understand how they relate. if we really need both we should definitely explain the difference in these files.

File: packages/flutter_tools/templates/app/analysis_options.yaml.tmpl (hunk 3 above)
Comment (goderbauer) [reply packages/flutter_tools/templates/app/analysis_options.yaml.tmpl:12]: I will fix this before submitting it. The Fix has already landed in the Dart SDK and engine, just waiting for a roll into the framework.

File: packages/flutter_tools/templates/app/pubspec.yaml.tmpl (hunk 7 above)
Comment (goderbauer) [reply packages/flutter_tools/templates/app/pubspec.yaml.tmpl:56]: Reworded the paragraph above. This line just makes the set of recommended lints available. The import in the `analysis_options.yaml` then activates that set.

**Review by pq — APPROVED** [review]
🚀

File: packages/flutter_tools/templates/app/lib/main.dart.tmpl [hunk 8]
```diff
@@ -113,14 +115,16 @@ class _MyHomePageState extends State<MyHomePage> {
...
     );
   }
 }
 {{/withPluginHook}}
 {{#withPluginHook}}
 class MyApp extends StatefulWidget {
+  const MyApp({Key? key}) : super(key: key);
+
   @override
   _MyAppState createState() => _MyAppState();
```
Comment (jonahwilliams) [inline packages/flutter_tools/templates/app/lib/main.dart.tmpl:129]: Does this conflict with the private return types lint? Or is that not in the recommended set

Comment (goderbauer) [reply packages/flutter_tools/templates/app/lib/main.dart.tmpl:129]: That's not in the recommended set.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Review by mit-mit — APPROVED** [review]

**Review by devoncarew — APPROVED** [review]

**Comment by goderbauer** [comment]
Thank you!

Submission is now blocked on the tree becoming green and https://github.com/dart-lang/linter/issues/2612 rolling into the framework.

**Comment by devoncarew** [comment]
> Thank you!
> 
> Submission is now blocked on the tree becoming green and [dart-lang/linter#2612](https://github.com/dart-lang/linter/issues/2612) rolling into the framework.

I believe the fix for this landed in dart-lang/sdk (https://github.com/dart-lang/sdk/commit/f193637d826113fa4ca074d1e075c8ee7ec42e62) ~8 days ago, so should likely already be available in flutter/flutter.

**Comment by goderbauer** [comment]
> I believe the fix for this landed in dart-lang/sdk (dart-lang/sdk@f193637) ~8 days ago, so should likely already be available in flutter/flutter.

While the fix has rolled into the engine, it's not available in flutter/flutter yet due to an unfortunate combination of engine rollbacks and tree closures. I believe the next engine roll into the framework will contain the fix, though, once the tree is open again.

My newly added tests is actually currently failing because the fix is not rolled in yet: https://github.com/flutter/flutter/pull/81417/checks?check_run_id=2520340935

**Comment by goderbauer** [comment]
Updated the version constraints to what https://dart-review.googlesource.com/c/sdk/+/198620 is doing.

**Comment by dnfield** [comment]
This broke skp_generator tests.

**Comment by goderbauer** [comment]
> This broke skp_generator tests.

Could parts of that test run presubmit (e.g. at least analyzing the code) to avoid this in the future?

**Comment by Hixie** [comment]
I don't see it there though, weird.

**Comment by Hixie** [comment]
It was added in https://github.com/flutter/flutter/pull/81096

**Comment by Hixie** [comment]
Fix in https://github.com/flutter/flutter/pull/82239

File: packages/flutter_tools/test/commands.shard/permeable/create_test.dart [hunk 9]
```diff
@@ -2567,11 +2615,27 @@ Future<void> _analyzeProject(String workingDir) async {
...
+      print(exec.stdout);
+      print(exec.stderr);
+    }
+    expect(exec.exitCode, 0);
+    return;
   }
-  expect(exec.exitCode, 0);
+  expect(exec.exitCode, isNot(0));
+  String lineParser(String line) {
+    print('#$line#');
```
Comment (Hixie) [inline packages/flutter_tools/test/commands.shard/permeable/create_test.dart:2628]: was this a temporary debugging line that leaked in?

Comment (goderbauer) [reply packages/flutter_tools/test/commands.shard/permeable/create_test.dart:2628]: Oh, yeah. Will send PR to remove.

Comment (goderbauer) [reply packages/flutter_tools/test/commands.shard/permeable/create_test.dart:2628]: https://github.com/flutter/flutter/pull/91317

**Pull request merged**